From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Radu Andrei Tudorica <raduandreitudorica3@gmail.com>
Date: Thu, 30 Jul 2026 20:30:00 +0300
Subject: [PATCH] Do not use __builtin_cpu_supports() for feature detection

bitops.c and hyperloglog.c call __builtin_cpu_supports() when HAVE_POPCNT
and HAVE_AVX2 are defined, which pulls in __cpu_model from libgcc. libgcc
is not linked into the unikernel, so the build fails at link time with
undefined references.

Leave both macros undefined so that the existing generic implementations
are used instead.

Signed-off-by: Radu Andrei Tudorica <raduandreitudorica3@gmail.com>
---
 src/config.h | 11 ++++++++---
 1 file changed, 8 insertions(+), 3 deletions(-)

diff --git a/src/config.h b/src/config.h
--- a/src/config.h
+++ b/src/config.h
@@ -330,7 +330,10 @@
 
 #if defined(__x86_64__) && ((defined(__GNUC__) && __GNUC__ > 5) || (defined(__clang__)))
     #if defined(__has_attribute) && __has_attribute(target)
-        #define HAVE_POPCNT
+        /* UNIKRAFT: __builtin_cpu_supports() requires __cpu_model from
+         * libgcc, which is not linked into the unikernel.  Leave
+         * HAVE_POPCNT undefined so that the software popcount is used.
+         */
         #define ATTRIBUTE_TARGET_POPCNT __attribute__((target("popcnt")))
     #else
         #define ATTRIBUTE_TARGET_POPCNT
@@ -342,7 +345,9 @@
 /* Check if we can compile AVX2 code */
 #if defined (__x86_64__) && ((defined(__GNUC__) && __GNUC__ >= 5) || (defined(__clang__) && __clang_major__ >= 4))
 #if defined(__has_attribute) && __has_attribute(target)
-#define HAVE_AVX2
+/* UNIKRAFT: see the HAVE_POPCNT comment above; leave HAVE_AVX2 undefined
+ * so that the generic HyperLogLog implementation is used.
+ */
 #endif
 #endif
 
